XTS/WS-T/dev/src/org/jboss/jbossts/xts/recovery/participant/ba/XTSBARecoveryManager.java

```diff
@@ -25,16 +25,19 @@ public abstract class XTSBARecoveryManager {
     public static XTSBARecoveryManager getRecoveryManager()
     {
         // use a lock to get the recovery manager to avoid a race with the code that sets it
-        synchronized (lock) {
             // theRecoveryManager must eventually be set to a non-null value
             while (theRecoveryManager == null) {
                 try {
-                    lock.wait();
+                    //syncrhronized block inside the loop for efficiency
+                    synchronized (lock) {
+
+                        lock.wait();
+                    
+                    }
```

Comment:
Could you please share your reasoning about this?

Reply:
Thanks for you reply @marcosgopen

> I think using the while loop inside the synchronized block is useless

Why? To me it seems to be very logical. Moreover, using a synchronized block inside a loop (as you changed) goes exactly against your comment `syncrhronized block inside the loop for efficiency`. Even though it might appear like a single instruction, acquiring a lock with `synchronized` is a very expensive operation and it shouldn't be used in a loop.

> ...**most of the time** the recoveryManager is already set.

That "most of the time" doesn't sound right. Synchronisation is used for concurrency -> "most of the time" doesn't cover when two (or more) processes try to modify the same variable.

> Additionally you might think that the synchronized block prevent other threads from changing the theRecoveryManager value, which is true inside the block, but actually the value can be changed by other threads before or after returning theRecoveryManager (https://github.com/jbosstm/narayana/blob/master/XTS/WS-T/dev/src/org/jboss/jbossts/xts/recovery/participant/ba/XTSBARecoveryManager.java#L40) because it is a reference to the object

I'm not sure I follow this but I guess it can be answered with my previous comment about "most of the time".

Reply:
Thanks again Manuel for your answer. I think this conversation is very helpful.
`acquiring a lock with synchronized is a very expensive operation and it shouldn't be used in a loop`
Correct me if I am wrong: from my understanding threads enter in that loop only once, where they wait until lock.notifyAll is invoked (https://github.com/jbosstm/narayana/blob/master/XTS/WS-T/dev/src/org/jboss/jbossts/xts/recovery/participant/ba/XTSBARecoveryManager.java#L59) and the while condition is not true anymore.
 `That "most of the time" doesn't sound right.`
 Sorry for that, I think this change created confusion. The change in XTSBARecoveryManager is not needed for a concurrency issue, this change is for efficiency only. This is why I said 'most of the time'.

Reply:
> from my understanding threads enter in that loop only once, where they wait until lock.notifyAll is invoked (https://github.com/jbosstm/narayana/blob/master/XTS/WS-T/dev/src/org/jboss/jbossts/xts/recovery/participant/ba/XTSBARecoveryManager.java#L59) and then the while condition (https://github.com/jbosstm/narayana/blob/master/XTS/WS-T/dev/src/org/jboss/jbossts/xts/recovery/participant/ba/XTSBARecoveryManager.java#L30) is not true anymore.

This "could" be true (i.e. it depends on many factors like compilation optimisations, the state of the variable in a precise moment, runtime optimisations, etc. ). It "is" true only if `synchronized` is used to protect the entire loop

Reply:
Mind I haven't studied how `theRecoveryManager` is used in any detail apart from reading the code. If you have a reasoning stronger than "efficiency", could you please share it (for my convenience)?